### PR TITLE
stat/distuv: Provide full test coverage for Gamma distribution and small fixes

### PR DESCRIPTION
**Pull request by romanwerpachowski**
- Full test coverage
- Removed redundant calculations in the small alpha version of Rand()
- Lowered the threshold for small alpha to 0.2 in accordance with https://www4.stat.ncsu.edu/~rmartin/Codes/rgamss.R
- Added the option to pass integration bounds to PDF normalisation check
- Other small fixes

Please take a look.

### CODE REVIEW

File: stat/distuv/distribution_test.go [hunk 1]
```diff
@@ -158,8 +158,12 @@ func checkQuantileCDFSurvival(t *testing.T, i int, xs []float64, c cumulanter, t
 }
 
 func checkProbContinuous(t *testing.T, i int, x []float64, p probLogprober, tol float64) {
-	// Check that the PDF is consistent (integrates to 1).
-	q := quad.Fixed(p.Prob, math.Inf(-1), math.Inf(1), 1000000, nil, 0)
+	checkProbContinuousWithBounds(t, i, x, math.Inf(-1), math.Inf(1), p, tol)
+}
+
+func checkProbContinuousWithBounds(t *testing.T, i int, x []float64, lb float64, ub float64, p probLogprober, tol float64) {
+	// Check that the PDF is consistent (integrates to 1). Takes lower and upper bounds for X.
```
Comment (kortschak) [inline stat/distuv/distribution_test.go:165]:
```
// checkProbContinuousWithBounds check that the PDF is consistent,
// it integrates to 1 from the lower to upper bound.
func checkProbContinuousWithBounds(t *testing.T, i int, x []float64, lower, upper float64, p probLogprober, tol float64) {
```

and associated param changes in the body.

If there is no `checkProbContinuous`, remove the `WithBounds` suffix. This is testing code, so the API doesn't need to be friendly and efficient, just obviously conprehensible.

File: stat/distuv/distribution_test.go [hunk 2]
```diff
@@ -261,7 +265,7 @@ func testRandLogProbContinuous(t *testing.T, i int, min float64, x []float64, f
 			return math.Exp(f.LogProb(x))
 		}
 		// Integrate the PDF to find the CDF
-		estCDF := quad.Fixed(prob, min, pt, 1000, nil, 0)
+		estCDF := quad.Fixed(prob, min, pt, 10000, nil, 0)
```
Comment (kortschak) [inline stat/distuv/distribution_test.go:268]: Why was this necessary?

File: stat/distuv/gamma.go [hunk 3]
```diff
@@ -93,6 +93,10 @@ func (g Gamma) Quantile(p float64) float64 {
 //
 // Rand panics if either alpha or beta is <= 0.
 func (g Gamma) Rand() float64 {
+	const (
+		// 0.2 threshold is from https://www4.stat.ncsu.edu/~rmartin/Codes/rgamss.R
```
Comment (kortschak) [inline stat/distuv/gamma.go:97]: The convention is fully formed sentences (also single line decl).

```
// The 0.2 threshold is from https://www4.stat.ncsu.edu/~rmartin/Codes/rgamss.R
// described in detail in https://arxiv.org/abs/1302.1884.
const smallAlphaThresh = 0.2
```

File: stat/distuv/gamma.go [hunk 4]
```diff
@@ -124,9 +128,8 @@ func (g Gamma) Rand() float64 {
 
 		// Algorithm adjusted to work in log space as much as possible.
 		lambda := 1/a - 1
-		lw := math.Log(a) - 1 - math.Log(1-a)
-		lr := -math.Log(1 + math.Exp(lw))
-		lc, _ := math.Lgamma(a + 1)
+		w := 1 / lambda / math.E
+		lr := -math.Log1p(w)
```
Comment (kortschak) [inline stat/distuv/gamma.go:132]: Please revert this. The original code intentionally works in log space (as described in the comment above).

File: stat/distuv/gamma.go [hunk 5]
```diff
@@ -135,18 +138,19 @@ func (g Gamma) Rand() float64 {
...
-				lEta = lc - z
+				lEta = -z
 			} else {
-				lEta = lc + lw + math.Log(lambda) + lambda*z
+				lEta = -1 + lambda*z
 			}
 			if lh-lEta > -exprnd() {
-				return math.Exp(-z/a) / b
+				return eza / b
 			}
```
Comment (kortschak) [inline stat/distuv/gamma.go:151]: Here too.

File: stat/distuv/gamma_test.go [hunk 6]
```diff
@@ -31,7 +31,9 @@ func TestGamma(t *testing.T) {
 	}
 	src := rand.New(rand.NewSource(1))
 	for i, g := range []Gamma{
+		{Alpha: 0.1, Beta: 0.8, Src: src},
```
Comment (kortschak) [inline stat/distuv/gamma_test.go:35]: Remove this blank line?

File: stat/distuv/distribution_test.go (hunk 1 above)
Comment (romanwerpachowski) [reply stat/distuv/distribution_test.go:165]: There is :) I could add the bounds to the original function, but it'd cause a cascade of changes everywhere.

File: stat/distuv/distribution_test.go (hunk 2 above)
Comment (romanwerpachowski) [reply stat/distuv/distribution_test.go:268]: PDFs with singularities (like Gamma for alpha < 1).

File: stat/distuv/gamma.go (hunk 4 above)
Comment (romanwerpachowski) [reply stat/distuv/gamma.go:132]: But it still does!

File: stat/distuv/distribution_test.go (hunk 1 above)
Comment (kortschak) [reply stat/distuv/distribution_test.go:165]: OK, just the signature/doc change then.

File: stat/distuv/distribution_test.go (hunk 2 above)
Comment (kortschak) [reply stat/distuv/distribution_test.go:268]: Thanks.

File: stat/distuv/gamma.go (hunk 5 above)
Comment (romanwerpachowski) [reply stat/distuv/gamma.go:151]: `lc` is unnecessary (we only ever need the difference `lh - lEta`).

Comment (kortschak) [reply stat/distuv/gamma.go:151]: Sorry, missed this. This is nice.

File: stat/distuv/gamma.go (hunk 4 above)
Comment (kortschak) [reply stat/distuv/gamma.go:132]: The question that is in my mind is whether `-math.Log1p(1 / lambda / math.E)` is better or worse than `-math.Log1p(math.Exp(-math.Log(lambda) - 1))`. I think you're probably right, that it's better. (They look to be about 1 ULP different in the end for reasonable small alpha.)

Given this, keep yours. Sorry.

**Comment by romanwerpachowski** [comment]
PTAL.

**Comment by kortschak** [comment]
You can go ahead and merge (please squash with the following commit message):

```
stat/distuv: provide full test coverage for Gamma distribution and small fixes

* Removed redundant calculations in the small alpha version of Rand.
* Lowered the threshold for small alpha to 0.2 in accordance with
  https://www4.stat.ncsu.edu/~rmartin/Codes/rgamss.R.
* Added the option to pass integration bounds to PDF normalisation check.
* Other small fixes.
```

**Comment by romanwerpachowski** [comment]
Thank you!

**Comment by kortschak** [comment]
For future, the single header line goes in the squash subject (no PR number link) and the body goes in the body. We try to match the convention used in the Go project.
